clippy_lints/src/ptr_eq.rs

```diff
@@ -0,0 +1,91 @@
+use crate::utils;
+use if_chain::if_chain;
+use rustc::hir::{BinOpKind, Expr, ExprKind};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+use rustc_errors::Applicability;
+
+declare_clippy_lint! {
+    /// **What it does:** Use `std::ptr::eq` when applicable
+    ///
+    /// **Why is this bad?** This can be used to compare `&T` references
+    /// (which coerce to `*const T` implicitly) by their address rather than
+    /// comparing the values they point to.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust
+    /// let a = &[1, 2, 3];
+    /// let b = &[1, 2, 3];
+    ///
+    /// let _ = a as *const _ as usize == b as *const _ as usize;
+    /// // Could be written
+    /// let _ = std::ptr::eq(a, b);
+    /// ```
+    pub PTR_EQ,
+    style,
+    "use `std::ptr::eq` when applicable"
+}
+
+declare_lint_pass!(PtrEqLint => [PTR_EQ]);
+
+static LINT_MSG: &str = "use `std::ptr::eq` when applicable";
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for PtrEqLint {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if let ExprKind::Binary(ref op, ref left, ref right) = expr.kind {
+            if BinOpKind::Eq == op.node {
+                let (left, right) = if_chain! {
+                    if let Some(lhs) = expr_as_cast_to_usize(cx, left);
+                    if let Some(rhs) = expr_as_cast_to_usize(cx, right);
+                    then {
+                        (lhs, rhs)
+                    }
+                    else {
+                        (&**left, &**right)
+                    }
+                };
+
+                if let Some(left_var) = expr_as_cast_to_raw_pointer(cx, left) {
+                    if let Some(right_var) = expr_as_cast_to_raw_pointer(cx, right) {
+                        let left_snip = utils::snippet(cx, left_var.span, "..");
+                        let right_snip = utils::snippet(cx, right_var.span, "..");
+                        let sugg = format!("std::ptr::eq({}, {})", left_snip, right_snip);
+                        utils::span_lint_and_sugg(
+                            cx,
+                            PTR_EQ,
+                            expr.span,
+                            LINT_MSG,
+                            "try",
+                            sugg,
+                            Applicability::MachineApplicable,
```

Comment:
How confident are you with the `MachineApplicable`ility? I'M asking because of the formulation of
```rust
static LINT_MSG: &str = "use `std::ptr::eq` when applicable";
```

Reply:
I've just added a macro check. But maybe there are more cases that I can't think of.